Source/Dafny/Compilers/Compiler-Csharp.cs

```diff
@@ -92,6 +92,62 @@ public CsharpCompiler(ErrorReporter reporter)
     }
 
     protected override void EmitBuiltInDecls(BuiltIns builtIns, ConcreteSyntaxTree wr) {
+      EmitInitNewArrays(builtIns, wr);
+
+      EmitFuncExtensions(builtIns, wr);
+    }
+
+    // Generates casts for functions of those arities present in the program, like:
+    //   public static class FuncExtensions {
+    //     public static Func<U, UResult> DowncastClone<T, TResult, U, UResult>(this Func<T, TResult> F,
+    //         Func<U, T> ArgConv, Func<TResult, UResult> ResConv) {
+    //       return arg => ResConv(F(ArgConv(arg)));
+    //     }
+    //     ...
+    //   }
+    // They aren't in any namespace to make them universally available.
+    private static void EmitFuncExtensions(BuiltIns builtIns, ConcreteSyntaxTree wr) {
+      var funcExtensions = wr.NewNamedBlock("public static class FuncExtensions");
+      foreach (var kv in builtIns.ArrowTypeDecls) {
+        int arity = kv.Key;
+
+        List<string> TyList(string prefix) {
+          var l = arity switch {
+            1 => new List<string> { prefix },
+            _ => Enumerable.Range(1, arity).Select(i => $"{prefix}{i}").ToList()
+          };
+          l.Add($"{prefix}Result");
+          return l;
+        }
+
+        string TPs(IEnumerable<string> l) => $"<{l.Comma()}>";
+
+        var us = TyList("U");
```

Comment:
It's a plural s, as in many Us.

Reply:
I'm ok with keeping it like this, and thanks for the name changes so far, but personally I wouldn't use variable names like this. I'm able to read this code because of the comment at the top of the method, so I know what the result should be and from that I can determine that every `u` relates to the output function and everything `t` relates to the input function, but I think ideally you'd be able to read the code without that comment.

I'd name this something like `inputFunctionTypes` and then indeed the interpolated string would become bigger: `public static Func<{inputFunctionTypes.Comma()}>`, but I'm OK with that.